### PR TITLE
Hubert

### PR DESCRIPTION
**Pull request by patrickvonplaten**
# What does this PR do?

This PR adds Hubert:

- https://ai.facebook.com/blog/hubert-self-supervised-representation-learning-for-speech-recognition-generation-and-compression
- https://arxiv.org/pdf/2106.07447.pdf?fbclid=IwAR3hI4uGqc4mV5j-ob8R5yLu-BaamVoe9ncxUoVmgFLjJXsE1IevP0rdNYY

Checkpoints are available here:
https://huggingface.co/models?filter=hubert

Hubert is essentially the same as Wav2Vec2 with some minor differences. The pretraining is completely different though, which is why we need to put it in a new modeling class. Pretraining functionality will be added in a second PR.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/master/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/master/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

File: src/transformers/models/hubert/modeling_hubert.py [hunk 1]
```diff
@@ -0,0 +1,1067 @@
...
+            return (input_length - kernel_size) // stride + 1
+
+        for kernel_size, stride in zip(self.config.conv_kernel, self.config.conv_stride):
+            input_lengths = _conv_out_length(input_lengths, kernel_size, stride)
+
+        return input_lengths
+
+
+HUBERT_START_DOCSTRING = r"""
+    Hubert was proposed in `wav2vec 2.0: A Framework for Self-Supervised Learning of Speech Representations
```
Comment (patrickvonplaten) [inline src/transformers/models/hubert/modeling_hubert.py:741]:
```suggestion
    Hubert was proposed in `HuBERT: Self-Supervised Speech Representation Learning by Masked Prediction of Hidden Units
```

File: src/transformers/models/hubert/modeling_hubert.py [hunk 2]
```diff
@@ -0,0 +1,1067 @@
...
+    <https://arxiv.org/abs/2006.11477>`__ by Alexei Baevski, Henry Zhou, Abdelrahman Mohamed, Michael Auli.
```
Comment (patrickvonplaten) [inline src/transformers/models/hubert/modeling_hubert.py:742]:
```suggestion
    <https://arxiv.org/abs/2106.07447>`__ by Wei-Ning Hsu, Benjamin Bolte, Yao-Hung Hubert Tsai, Kushal Lakhotia, Ruslan Salakhutdinov, Abdelrahman Mohamed.
```

File: src/transformers/models/hubert/configuration_hubert.py [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+
+    Configuration objects inherit from :class:`~transformers.PretrainedConfig` and can be used to control the model
+    outputs. Read the documentation from :class:`~transformers.PretrainedConfig` for more information.
+
+
+    Args:
+        vocab_size (:obj:`int`, `optional`, defaults to 32):
+            Vocabulary size of the Hubert model. Defines the number of different tokens that can be represented by the
+            :obj:`inputs_ids` passed when calling :class:`~transformers.HubertModel` or
+            :class:`~transformers.TFHubertModel`. Vocabulary size of the model. Defines the different tokens that can
```
Comment (patil-suraj) [inline src/transformers/models/hubert/configuration_hubert.py:44]: there's no `TFHubertModel` yet no?

File: tests/test_configuration_auto.py [hunk 4]
```diff
@@ -50,4 +50,5 @@ def test_pattern_matching_fallback(self):
         # no key string should be included in a later key string (typical failure case)
         keys = list(CONFIG_MAPPING.keys())
         for i, key in enumerate(keys):
+            print(key)
```
Comment (LysandreJik) [inline tests/test_configuration_auto.py:53]: This can be removed

File: src/transformers/models/hubert/modeling_hubert.py [hunk 5]
```diff
@@ -0,0 +1,1065 @@
...
+            >>> from datasets import load_dataset
+            >>> import soundfile as sf
+
+            >>> processor = HubertProcessor.from_pretrained("facebook/wav2vec2-base-960h")
+            >>> model = HubertModel.from_pretrained("facebook/wav2vec2-base-960h")
+
+            >>> def map_to_array(batch):
+            >>>     speech, _ = sf.read(batch["file"])
+            >>>     batch["speech"] = speech
+            >>>     return batch
```
Comment (LysandreJik) [inline src/transformers/models/hubert/modeling_hubert.py:881]:
```suggestion
            ...     speech, _ = sf.read(batch["file"])
            ...     batch["speech"] = speech
            ...     return batch
```

File: src/transformers/models/hubert/modeling_hubert.py [hunk 6]
```diff
@@ -0,0 +1,1065 @@
...
+            >>> from datasets import load_dataset
+            >>> import soundfile as sf
+
+            >>> processor = HubertProcessor.from_pretrained("facebook/wav2vec2-base-960h")
+            >>> model = HubertForCTC.from_pretrained("facebook/wav2vec2-base-960h")
+
+            >>> def map_to_array(batch):
+            >>>     speech, _ = sf.read(batch["file"])
+            >>>     batch["speech"] = speech
+            >>>     return batch
```
Comment (LysandreJik) [inline src/transformers/models/hubert/modeling_hubert.py:994]:
```suggestion
            ...     speech, _ = sf.read(batch["file"])
            ...     batch["speech"] = speech
            ...     return batch
```

File: src/transformers/models/hubert/modeling_hubert.py [hunk 7]
```diff
@@ -0,0 +1,1065 @@
...
+            >>> predicted_ids = torch.argmax(logits, dim=-1)
+
+            >>> transcription = processor.decode(predicted_ids[0])
+
+            >>> # compute loss
+            >>> target_transcription = "A MAN SAID TO THE UNIVERSE SIR I EXIST"
+
+            >>> # wrap processor as target processor to encode labels
+            >>> with processor.as_target_processor():
+            >>>     labels = processor(target_transcription, return_tensors="pt").input_ids
```
Comment (LysandreJik) [inline src/transformers/models/hubert/modeling_hubert.py:1010]:
```suggestion
            ...     labels = processor(target_transcription, return_tensors="pt").input_ids
```

**Review by LysandreJik — APPROVED** [review]
LGTM! Very simple model, cool that  it doesn't use a feature processor/tokenizer.

Shouldn't a feature processor be added for batching, padding, truncating, etc, and for decoding rather than relying on the `Wav2Vec2CTCTokenizer`?

File: src/transformers/models/hubert/modeling_hubert.py [hunk 8]
```diff
@@ -0,0 +1,1065 @@
...
+
+    def freeze_feature_extractor(self):
+        """
+        Calling this function will disable the gradient computation for the feature extractor so that its parameter
+        will not be updated during training.
+        """
+        self.hubert.feature_extractor._freeze_parameters()
+
+    @add_start_docstrings_to_model_forward(HUBERT_INPUTS_DOCSTRING)
+    @replace_return_docstrings(output_type=BaseModelOutput, config_class=_CONFIG_FOR_DOC)
```
Comment (patil-suraj) [inline src/transformers/models/hubert/modeling_hubert.py:962]:
```suggestion
    @replace_return_docstrings(output_type=CausalLMOutput, config_class=_CONFIG_FOR_DOC)
```

File: tests/test_configuration_auto.py (hunk 4 above)
Comment (patil-suraj) [inline tests/test_configuration_auto.py:53]: Think this is leftover from debugging

**Comment by patrickvonplaten** [comment]
> batching

It uses `Wav2Vec2Processor` for feature extraction etc

**Review by patil-suraj — APPROVED** [review]
Looking good!
